### ISSUE TITLE
Review readiness of Auto Login URL module for Drupal 10

### ISSUE THREAD

**Issue by ultimike** [communityinclusion/escoach#107]
https://www.drupal.org/project/auto_login_url

This might be useful for our "login by URL" task in phase 2.

-mike

**Comment by ultimike** [comment]
The Auto login url module is good to go with this patch/MR: https://www.drupal.org/project/auto_login_url/issues/3303974

We’ll probably have to use https://github.com/mglaman/composer-drupal-lenient to install the D9 version of the module on a D10 site, as well as Composer patches to apply the patch.

The way the module uses permissions is a bit nonsensical (IMHO), so be sure to read my comment here: https://www.drupal.org/project/auto_login_url/issues/2878078#comment-15284108

I tested it with querystring variables in the destination - appears to work fine.

In code, to create an autologin URL for UID=54 (for example): `\Drupal::service('auto_login_url.create')->create(54, 'some/path?blah=34325');`

-mike

**Comment by ultimike** [comment]
@thinkcollege Exactly how do you envision using these login URLs? Do you want to be able to click a link on a user profile to generate a login link for that user? 

How long should the link be good for before it expires? (less is **much** better from a security perspective - I suggest less than 7 days). 

-mike

**Comment by thinkcollege** [comment]
@ultimike  It would be great if we could generate a spreadsheet of all users with, say, names, phone numbers, email and the unique login link. I agree, the links should expire.  I told Alberto they should expire after 1 day.  He wanted two weeks.  I will make an executive decision and say 5 days is all we need.  Could we set up a system where when you manually generate the spreadsheet with all the links the old ones expire, whether the time limit has been reached or not? I think the best use of the link will be in the daily survey, so if we use the link on two consecutive days we will want to cancel the previous one, right?

**Comment by ultimike** [comment]
@thinkcollege I believe I understand what you're asking, but allow me to restate in a different form:

1. From the admin menu's Report section, provide a "Download users" link. When this link is clicked:
2. All previous Auto Login Urls are cancelled. 
3. All active site users are output to a .csv file that contains first name, last name, phone number, email address, and unique login link. 
4. The unique login link is valid for 5 days from the time of creation. 

Is all this correct?

Where should the unique login link lead to? The home page?

-mike

**Comment by thinkcollege** [comment]
Talking with Alberto about this, I have some other thoughts.  On your point 3 the login link should just be the login slug from the URL.  That way we could also insert it into a customized link to, say, the legacy dashboard if we wanted to.
Another big ask:  That login slug should be automatically inserted into a field in the user's profile.  We want to insert those login links in the daily survey as needed.  Our Alchemer survey pulls fields from user profiles, so I can set up a field in the survey that would pull from that user profile field.  I would have to create a compound field in the daily survey that appended the slug to whatever custom ES-Coach URL we wanted the survey participants to go to.  But that is my problem, I guess.  It would be good if the login id went at the end of the URL query string to facilitate appending it to another field with the custom link. 
To sum up: login slug to be generated on demand by admins in a spreadsheet and also get inserted in a field in the user profile.

**Comment by ultimike** [comment]
@thinkcollege Unfortunately, I don't think what you're asking is within our budget. The way the Auto Login Url module works is that the login destination is needed at the time that the hash value is created - and it becomes part of the hash. So, it is not possible to first generate the login hash, then attach a destination to it.

If the destination URL is the same for all users, or would it be different depending on the user? If the former then perhaps as part of the process of generating the .csv file, we could prompt for the destination. 

If the Auto Login Url is `automatically inserted into a field in the user's profile` - when would this happen? Once per day via cron? 

I would really like to start with a minimum viable product for this task and then add complexity as time and budget allows.

-mike

**Comment by thinkcollege** [comment]
The URL would be the same for all users on a given day or for a given instance, so would it be possible to have a field to type the generic URL before generating the unique hashed URLs for all users?  
I would prefer that the login urls are generated on demand by the administrator rather than by cron.  
I would like the URLs inserted in a user profile field (as well as the spreadsheet).  That would be by far the best way to get them into the daily survey.

**Comment by ultimike** [comment]
Based on the last few comments, how about this:

1.  From the admin menu's Report section, provide a "Generate Auto Login Urls" link.
2. When the link is clicked, the user is presented with a form where they can enter the URL for where they'd like **all** users redirected once they click their Auto Login Url link. 
3. When the form is submitted, all previous Auto Login Urls are cancelled AND
4. All active site users are output to a .csv file that contains first name, last name, phone number, email address, and unique login link that has the destination URL embedded AND
5. The unique login link is valid for 5 days from the time of creation AND
6. All user entities have their individual Auto Login Url links saved to a field to be determined AND
7. Either the .csv file is automatically downloaded or a download link is provided.

Am I missing anything?

-mike

**Comment by thinkcollege** [comment]
That sounds like a plan.  Go ahead.

**Comment by ultimike** [comment]
@thinkcollege For item 6 above, where would you like the Auto Login Url field to be created? Directly on the user entity? On one of the user profiles (`survey_participants`, perhaps?) 

Do you have a preference for what the field is called?

-mike

**Comment by mgcobblestone** [comment]
@ultimike , @thinkcollege I think it would be best to add the field to the survey_participants profile.

Also, can we assume that links should only be generated for users that have a Job Type = Employment Consultant ( and maybe both manager/consultant) ?

One last question, do we want the links to be single-use or reusable?

**Comment by thinkcollege** [comment]
Yes, the field with the auto-login URL should be on the survey_participants profile.  Name could be "auto_login_url."  Name is not a big issue because I will hide it from the profile owners/users.  It is only for purposes of feeding the url to the Alchemer survey API. I think the links could be reusable during the 5-day validity period, or until another set of URLs is generated.

**Comment by ultimike** [comment]
@thinkcollege This is ready for your testing. Before you pull (`es_home` branch), be sure to export and commit any new config that you have made. Then, after you pull, do the following:

1. Run `composer install` to get Auto Login Url module. 
2. Enable Auto Login Url module. 
3. Do a config import. 
4. Rebuild caches. 
5. From the admin menu, in "Reports", select "Export users with auto login links" - the rest should be self-explanatory.

Let us know how it goes.

thanks,
-mike

**Comment by thinkcollege** [comment]
@ultimike I tried this out on my local and on rp-escoach.communityinclusion.org  Same result on both:

1. A spreadsheet is generated with name, phone, email and login link.  The list of users on the spreadsheet is partial, not complete, seemingly random, including .  Are there any criteria set for inclusion in the csv?  Why are some (most)users not included?  The included users with links in the list should be all registered site users.
2. The links are filled in in the auto login url profile field, but only for the limited list of users in the CSV file.
3. The login link works correctly, taking the user to the correct page, and logging them in.

**Comment by mgcobblestone** [comment]
@thinkcollege See my question from 11/10.  The list is currently filtered as I had proposed, though I realize now I should have also included managers.  Let me know if that is correct or if other users should be included as well.

**Comment by thinkcollege** [comment]
@mgcobblestone  The list for the URLs should include all Active users on the site.  I guess I didn't reply to your question. That should cover all bases in case we add new roles, or users have multiple roles and so on.  It doesn't hurt anything to be thorough, and nothing is getting sent out based on the creation of the URLs.

**Comment by mgcobblestone** [comment]
Thanks @thinkcollege , I have pushed an update that removes the filtering.  Keep in mind that all of the contact fields are coming from the survey participant profiles so only users that have a profile will be included.

One other note:  Currently, ALL previous login links will be deleted when the form is submitted to generate a new set of links.  I can either put a warning on the form with a message to that effect, or I can add a checkbox making that behavior an option. Or I can leave it as-is.  Let me know which you prefer.

**Comment by thinkcollege** [comment]
Looks good without the filtering.  All users with profiles show up in the CSV and are getting a profile entry. I like the idea of a checkbox to retain previous links.  Assuming they will expire after 5 days.

**Comment by ultimike** [comment]
Assigning to @mgcobblestone for the addition of the checkbox (see previous comment).

-mike

**Comment by thinkcollege** [comment]
@mgcobblestone @ultimike  I am just realizing that the Auto Login module permanently changes the password of users, including admins.  I hadn't taken that into account, nor thought of how this affects people who have saved their passwords in browsers or password managers. I guess we need to filter on Admins, at least, and we still need to think about how this affects our user policies.

**Comment by ultimike** [comment]
@thinkcollege I do not believe the code of the Auto Login Url module changes any passwords. I just re-reviewed some of the code of the module and I don't see any evidence of that.

Can you provide steps to demonstrate where you feel this is happening?

thanks,
-mike

**Comment by thinkcollege** [comment]
@ultimike @mgcobblestone Let me look into this more.  It was happening on my local, with my root account only.  I see it is not happening to other accounts and not on the development site.  Hold off on doing any filtering or anything until I see what is up.

**Comment by thinkcollege** [comment]
@ultimike @mgcobblestone  Never mind my last comments.  Auto Login URL is working okay.  Had some very weird behavior where my root user was getting logged out by the module.  May have been a local database issue. I was able to fix it.

**Comment by mgcobblestone** [comment]
@thinkcollege New commit is up with checkbox option for deleting previous link.

cc: @ultimike

**Comment by thinkcollege** [comment]
@ultimike @mgcobblestone Could we add the profile field Job Type to the spreadsheet with the auto logins?  That would help in sorting them for mailings.  Thanks.

**Comment by mgcobblestone** [comment]
@thinkcollege New commit is up with Job Type field in export

cc: @ultimike

**Comment by thinkcollege** [comment]
@ultimike @mgcobblestone  
It would really help me if there were an ability to clear the autologin URL profile field.  This functionality would be ideally, on the Autologin URL admin page where you set a new URL.  I would like the option to clear the profile field without setting a new URL, with a checkbox and submit.  That checkbox could be labeled "Clear all URLs and clear the Autologin URL field in profile without setting a new URL". Rationale: I am using Alchemer's API to pass the URL to the daily survey.  I never want there to be a case where an expired URL is in place and gets passed to the survey.  If we could clear this field when not needed I could have the API disregard an empty field and not pass along the value

**Comment by mgcobblestone** [comment]
Hi @thinkcollege ,
I have pushed an update. It deletes **ALL** auto-login URLs and clears the auto login field from **ALL** profile records.

cc: @ultimike

**Comment by thinkcollege** [comment]
@mgcobblestone @ultimike  This is working as expected.
